pkg/scalers/azure_monitor.go

```diff
@@ -0,0 +1,206 @@
+	Filter                    string
+	ResourceGroup             string
+	Topic                     string
+	Subscription              string
+}
+
+type azureExternalMetricResponse struct {
+	Value float64
+}
+
+type azureExternalMetricClient interface {
+	getAzureMetric(azMetricRequest azureExternalMetricRequest) (azureExternalMetricResponse, error)
+}
+
+type insightsmonitorClient interface {
+	List(ctx context.Context, resourceURI string, timespan string, interval *string, metricnames string, aggregation string, top *int32, orderby string, filter string, resultType insights.ResultType, metricnamespace string) (result insights.Response, err error)
+}
+
+type monitorClient struct {
+	client insightsmonitorClient
+}
+
+// GetAzureMetricValue is a func
+func GetAzureMetricValue(ctx context.Context, metricMetadata *azureMonitorMetadata) (int32, error) {
+	metricsClient := newMonitorClient(metricMetadata)
+
+	metricRequest := azureExternalMetricRequest{
+		Timespan:       timeSpan(),
+		SubscriptionID: metricMetadata.subscriptionID,
+	}
+
+	metricRequest.MetricName = metricMetadata.name
+	metricRequest.ResourceGroup = metricMetadata.resourceGroupName
+	resourceInfo := strings.Split(metricMetadata.resourceURI, "/")
+
+	if len(resourceInfo) != 3 {
+		return -1, fmt.Errorf("resourceURI is missing resource namespace, resource type, or resource name")
+	}
+
+	metricRequest.ResourceProviderNamespace = resourceInfo[0]
+	metricRequest.ResourceType = resourceInfo[1]
+	metricRequest.ResourceName = resourceInfo[2]
+
+	metricRequest.Aggregation = metricMetadata.aggregationType
+
+	filter := metricMetadata.filter
+	if filter != "" {
+		metricRequest.Filter = filter
+	}
+
+	aggregationInterval := metricMetadata.aggregationInterval
+	if aggregationInterval != "" {
+		metricRequest.Timespan = aggregationInterval
+	}
+
+	metricResponse, err := metricsClient.getAzureMetric(metricRequest)
+	if err != nil {
+		azureMonitorLog.Error(err, "error getting azure monitor metric")
+		//return -1, fmt.Errorf("MetricName %s: , ResourceGroup: %s, Namespace: %s, ResourceType: %s, ResourceName: %s, Aggregation: %s, Timespan: %s", metricRequest.MetricName, metricRequest.ResourceGroup, metricRequest.ResourceProviderNamespace, metricRequest.ResourceType, metricRequest.ResourceName, metricRequest.Aggregation, metricRequest.Timespan)
+		return -1, fmt.Errorf("Error getting azure monitor metric %s: %s", metricRequest.MetricName, err.Error())
+	}
+
+	// casting drops everything after decimal, so round first
+	metricValue := int32(math.Round(metricResponse.Value))
+
+	return metricValue, nil
+}
+
+func newMonitorClient(metadata *azureMonitorMetadata) azureExternalMetricClient {
+	client := insights.NewMetricsClient(metadata.subscriptionID)
+	config := auth.NewClientCredentialsConfig(metadata.clientID, metadata.clientPassword, metadata.tentantID)
+
+	authorizer, err := config.Authorizer()
+	if err == nil {
+		client.Authorizer = authorizer
+	}
+
+	return &monitorClient{
+		client: client,
```

Comment:
Is there a number of different clients Azure would give? Do they not have a public facing interface we can use? If we can avoid needing to maintain this interface in perpetuity would be better in the long term.

Reply:
Let me take a look at this. [Here is the relevant code from the sdk](https://github.com/Azure/azure-sdk-for-go/blob/master/services/preview/monitor/mgmt/2018-03-01/insights/metrics.go). I need to play around with it, but my initial thought is that yes, we could use the public facing interface. The only possible problem is that the `monitorClient` is used as a receiver in `getAzureMetric`, but I could just pass in `monitorClient` to the function instead. Thoughts?

Reply:
in terms of code organization, I think I'd have 1 method `getAzureMetricValue` that's private, (it doesn't have to be public, right?) and split the code you have in it into `createMetricsClient(metricMetadata)`, `createMetricsRequest(metricMetadata)`, `executeRequest(client, request)`?

Reply:
`GetAzureMetricValue` is called in a different file so I assumed it needed to be public? But I'm not a golang expert. But in terms of splitting the code, I like that idea and will change the code accordingly.

Reply:
You're right I did think it was in the same file when I said that. although in go it can still be private because they are both in the same package, I see all other scalers have those methods public, so it's fine as it's. Sorry about the confusion.